crates/uv/src/lib.rs

```diff
@@ -42,8 +44,50 @@ pub(crate) mod printer;
 pub(crate) mod settings;
 pub(crate) mod version;
 
+/// Resolves the script target for a run command.
+///
+/// If it's a local file, does nothing. If it's a URL, downloads the content
+/// to a temporary file and updates the command. Prioritizes local files over URLs.
+/// Returns Some(NamedTempFile) if a remote script was downloaded, None otherwise.
+async fn resolve_script_target(
+    command: &mut ExternalCommand,
+) -> Result<Option<tempfile::NamedTempFile>> {
+    let Some(target) = command.get_mut(0) else {
+        return Ok(None);
+    };
+
+    // Only continue if we are absolutely certain no local file exists.
+    if let Ok(false) = Path::new(target).try_exists() {
+    } else {
+        return Ok(None);
+    }
+
+    let maybe_url = target.to_string_lossy();
+
+    // Only continue if the target truly looks like a URL.
+    if !(maybe_url.starts_with("http://") || maybe_url.starts_with("https://")) {
+        return Ok(None);
+    }
+
+    let script_url = url::Url::parse(&maybe_url)?;
+    let file_name = script_url
+        .path_segments()
+        .and_then(std::iter::Iterator::last)
+        .unwrap_or("script.py");
+
+    let temp_file = tempfile::Builder::new()
+        .prefix(file_name)
+        .suffix(".py")
```

Comment:
Will this result in `script.py.py` if the `script.py` fallback is used above?

Reply:
Ah yeah, you are right. tempfile adds a hash and a suffix to the prefix.

```sh
cargo run run https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py CI
# Reading inline script metadata from: /var/folders/gs/j5y7yv816z71dzy18r4qmy3h0000gq/T/uv-run-remote-script-test.pyznxLtC.py
# Installed 4 packages in 21ms
# Hello CI, from uv!
```

It might be easiest to:

```rust
    let temp_file = tempfile::Builder::new()
        .suffix(".py")
        .tempfile?;
```

Depends on how useful it is to try for the file name. It's only used to report to the end user and would show up in traceback messages from Python errors.

Reply:
I think including the file name in there as a prefix is fine.  Especially if we add a log message saying that the remote script was written to the temp file.

I think my issue here is the case where the `script.py` _fallback name_ is used (from above). Does that result in a double `.py` suffix?

Reply:
I think the issue also happens if the file has as `.py`. For example:

https://raw.githubusercontent.com/astral-sh/uv/main/scripts/uv-run-remote-script-test.py

results in:

uv-run-remote-script-test.pyznxLtC.py

So `script.py` would similarly result in `script.py[hash].py`.

Reply:
Yeah, so I think you want to strip the `.py` suffix from the file name, but keep the `.py` suffix in the temp file builder.

Reply:
Ah, much more simple. Done!